simulator.py

```diff
@@ -0,0 +1,96 @@
+"""
+Simulator that simulates trading process and estimates profits and losses.
+"""
+
+import copy
+import uuid
+from thousandaire.dataset import Data
+
+DATASET = 'price_dataset'
+
+def decode_data(data):
+    """
+    Decode while in multi-process.
+
+    In multi-processing,
+    the simulator will get data by shared memory,
+    which are encoded by pickle to make them sharable.
+    To avoid polluting the shared memory, we need to create a local copy.
+    Since 'target_data' may also in 'others', that we should deepcopy the data
+    separately not simply using copy.deepcopy(data).
+    """
+    return {
+        'workdays' : copy.deepcopy(data['workdays']),
+        'target_data' : copy.deepcopy(data['target_data']),
+        'others' : copy.deepcopy(data['others'])}
+
+class Simulator:
+    """
+    Simulate the alpha.
+    """
+    def __init__(self, settings, data, cost_function):
+        self.cost_function = cost_function
+        self.data = decode_data(data)
+        self.settings = settings
+        self.__result = Data(
+            'result', ['pnl', 'cost', 'positions_raw', 'positions_np'])
+        self.__portfolio = list()
+        self.__key = uuid.uuid4()
+        self.initialize_data()
+
+    def generate_pnl(self, date):
+        """
+        Calculate pnl for alphas.
+        This method will be called by simulation day by day while inputting
+        a new portfolio.
+
+        Profit and cost will return by user-specified cost_function.
+        """
+        profit, cost = self.cost_function(
+            self.__portfolio, self.data['target_data'])
+        self.__result.append(
+            (date, profit, cost, self.__portfolio[-1],
+             self.__portfolio[-1].encode_to_nparray(self.settings.target)))
+
+    def next_date(self):
+        """
+        Move the simulating date.
+        """
+        for dataset in self.data['others'].values():
+            dataset.next_date(self.__key)
+        self.data['target_data'].next_date(self.__key)
+        self.data['workdays'].next_date(auth_key=self.__key)
+
+    def run(self):
+        """
+        Carried out simulating process.
+        """
+        alpha_formula = self.settings.alpha(
+            self.settings.start_date, self.data['others'],
+            self.settings.parameters)
+        while (self.data['workdays'].get_today() is None or
+               self.data['workdays'].get_today() <= self.settings.end_date):
+            portfolio = alpha_formula(
+                self.data['workdays'].get_today(), self.data['others'])
+            portfolio.normalize(self.data['workdays'].get_today())

```

Comment:
Actually we can handle the "zero portfolio exception" here.
Do nothing in Portfolio.normalilze, but handle the exception here.